src/plugins/deadlines/index.tsx

```diff
@@ -0,0 +1,258 @@
+      }
+    }
+  }
+
+  private async inDeadlines(row: Row) {
+    // check if row is in top level of deadlines
+    this.log('inDeadlines', row);
+    const root = await this.getDeadlinesRoot();
+    const children = await this.getChildren(root);
+    for (const child of children) {
+      if (child.row === row) {
+        return true;
+      }
+    }
+    return false;
+  }
+
+  private async createDeadlineClone(row: Row, thisDate: Date) {
+    this.log('createDeadlineClone', row, thisDate);
+    const root = await this.getDeadlinesRoot();
+    let index = 0;
+    const children = await this.getChildren(root);
+    for (const child of children) {
+      const date = await this.getDate(child);
+      if (date && thisDate < date) {
+        break;
+      }
+      index++;
+    }
+    this.log('createDeadlineClone attachBlocks', root, row, index);
+
+    await this.api.session.attachBlocks(root, [row], index);
+    await this.api.updatedDataForRender(row);
+  }
+
+  private async getDate(path: Path) {
+    // get date of a deadline
+    this.log('getDate', path);
+    const children = await this.getChildren(path);
+    for (const child of children) {
+      const text = await this.api.session.document.getText(child.row);
+      const date = await this.parseDate(text);
+      if (date) {
+        return date;
+      }
+    }
+    return null;
+  }
+
+  private async parseDate(text: string) {
+    // return Date or null if invalid
+    this.log('parseDate', text);
+    return this.parseFullDate(text);  // might add more parse options
+  }
+
+  private async parseFullDate(text: string) {
+    this.log('parseFullDate', text);
+    const regex = matchWordRegex('due ((\\d+\\-)?\\d+\\-\\d+)');
+    let match = regex.exec(text.toLowerCase());
+    if (match) {
+      this.log('Matched', match);
+      const d = match[2].split('-').map((Number));
+      let dateStr = match[2];
+      const today = new Date();
+      if (d.length === 2) {
+        dateStr = today.getFullYear().toString() + '-' + dateStr;
+      }
+      const date = new Date(Date.parse(dateStr));
+      const lastWeek = new Date(today);
+      lastWeek.setDate(today.getDate() - 7);
+      if (d.length === 2 && date < lastWeek) {
+        date.setFullYear(date.getFullYear() + 1);
+      }
+      this.log('parseFullDate', d, date);
+      return date;
+    }
+    return null;
+  }
+
+  private async checkDeleted(info: CachedRowInfo) {
+    this.log('checkDeleted', info);
+    let needReInit = false;
+    const root = await this.getDeadlinesRoot();
+    if (root) {
+      if (info.row === root.row) {
+        needReInit = true;
+      }
+    }
+    if (needReInit) {
+      this.log('needReInit', true);
+      this.deadlinesRoot = null;
+      await this.init();
+    }
+  }
+
+  public async log(...args: any[]) {
+    if (this.isLogging) {
+      this.logger.info('Deadlines: ', ...args);
+    }
+  }
+
+  private async getNodeWithText(root: Path, text: String): Promise<Path | null> {
```

Comment:
for efficiency, you could consider using the Marks feature

Reply:
I dont think its worth it to change it since this function is used in other places for error checking. Nodes probably don't have many children.

Reply:
it's only ever used to search for "deadlines" right?

some other benefits:  
- you can have the deadlines node wherever you want, instead of at the document root
- uniqueness is guaranteed

Reply:
Yeah you're right. I'm having trouble figuring out how to check if the mark exists. I see theres a public method listMarks but I don't know how to call it.

Reply:
ahh, yeah, sorry!  I think I need to add something to PluginAPI so that you can access other plugins.  let me try briefly, the daily notes plugin has a similar issue

Reply:
i added a way to do this, see https://github.com/WuTheFWasThat/vimflowy/commit/bb6c81eb9320247e8747727a07462394a347f6ea

let me know if you have questions, it's totally new so not very tested